Modules/Devices/src/Devices.Application/Tiers/Commands/DeleteTier/Handler.cs

```diff
@@ -0,0 +1,52 @@
+﻿using Backbone.Modules.Devices.Application.Infrastructure.Persistence.Repository;
+using Backbone.Modules.Devices.Application.IntegrationEvents.Outgoing;
+using Backbone.Modules.Devices.Domain.Aggregates.Tier;
+using Enmeshed.BuildingBlocks.Application.Abstractions.Infrastructure.EventBus;
+using MediatR;
+using ApplicationException = Enmeshed.BuildingBlocks.Application.Abstractions.Exceptions.ApplicationException;
+
+namespace Backbone.Modules.Devices.Application.Tiers.Commands.DeleteTier;
+
+public class Handler : IRequestHandler<DeleteTierCommand>
+{
+    private readonly ITiersRepository _tiersRepository;
+    private readonly IEventBus _eventBus;
+
+    public Handler(ITiersRepository tiersRepository, IEventBus eventBus)
+    {
+        _tiersRepository = tiersRepository;
+        _eventBus = eventBus;
+    }
+
+    public async Task Handle(DeleteTierCommand request, CancellationToken cancellationToken)
+    {
+        TierId tierId;
+
+        try
+        {
+            tierId = TierId.Create(request.TierId).Value;
+        }
+        catch (Exception)
+        {
+            throw new ApplicationException(ApplicationErrors.Devices.InvalidTierId());
+        }
+
+        var tier = await _tiersRepository.FindById(tierId, cancellationToken);
+
+        if (tier.IsBasicTier())
+        {
+            throw new ApplicationException(ApplicationErrors.Devices.BasicTierCannotBeDeleted());
+        }
+
+        var identitiesCount = await _tiersRepository.GetIdentitiesCount(tier, cancellationToken);
+
+        if (identitiesCount > 0)
+        {
+            throw new ApplicationException(ApplicationErrors.Devices.UsedTierCannotBeDeleted(identitiesCount));
+        }
```

Comment:
You should do this in a more domain centric way:
```suggestion
        var identitiesCount = await _tiersRepository.GetIdentitiesCount(tier, cancellationToken);

        var impediment = tier.CanBeDeleted(identitiesCount)
        if(impediment != null)
        {
            throw new DomainException(impediment);
        }
```

The `CanBeDeleted` method then checks for the Name and the number of identities and returns a DomainError if there is something wrong. If everything is fine, it returns null.

Regarding the tests:
* you should write tests for the `CanBeDeleted` method. 
* you can simplify the HandlerTests by boiling them down to only a single test that checks for an exception. You can simulate the exception by returning an invalid Tier from the fake repository.

Reply:
Regarding the tests, do you mean we can move the exception-specific logic from the handler tests to the `CanBeDeleted` tests?

Reply:
Yep. Except that the CanBeDeleted method won't throw exceptions but instead return a DomainError if there is something wrong